### ISSUE TITLE
Feature: Implement Default Parameters Requirements

### ISSUE THREAD

**Issue by DiogenesAnalytics** [DiogenesAnalytics/autoencoder#10]
# Problem
Currently the only *"requirements"* of the the `BaseLayerParams.default_parameters` property is that it is of type `DefaultParams`: https://github.com/DiogenesAnalytics/autoencoder/blob/ca2c35b185c0c106b282d6bea6077aa7ba04c3cf/src/autoencoder/model/base.py#L16-L62

But there will likely be a need for even more *intense* requirements of this *data structure* (i.e. in order to determine which layers are `input`, `encoding`, `decoding`, see issue #17 for a this specific feature).

## References
+ https://stackoverflow.com/questions/36849837/introspecting-arguments-from-the-constructor-function-init-in-python
+ https://docs.python.org/3.9/library/inspect.html#introspecting-callables-with-the-signature-object
+ https://docs.python.org/3.9/library/exceptions.html

**Comment by DiogenesAnalytics** [comment]
# Solution
This can be *resolved* by simply *flipping* the *instance dict* with the *default parameters dict*:
```python
# loop over layer_name/params in defaults
for layer_id, params in self.default_parameters.items():
    # now find corresponding layer_id in constructor args
    try:
        yield layer_id, self.__dict__[layer_id]
    except KeyError as error:
        raise KeyError(
            "Constructor arguments must match "
            "the keys of the default_parameters dictionary."
        ) from error
```

**Comment by DiogenesAnalytics** [comment]
Implemented with: 3282476194d414459a9762f879e744d29521a82c
